cirq-core/cirq/optimizers/two_qubit_to_sqrt_iswap.py

```diff
@@ -0,0 +1,413 @@
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    return True
+
+
+def _decomp_0_matrices(
+    kak: 'cirq.KakDecomposition',
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Returns the single-qubit matrices for the 0-SQRT_ISWAP decomposition.
+
+    Assumes canonical x, y, z and (x, y, z) = (0, 0, 0) within tolerance.
+    """
+    # Pairs of single-qubit unitaries, SQRT_ISWAP between each is implied
+    # Only a single pair of single-qubit unitaries is returned here so
+    # _decomp_to_operations will not insert any sqrt-iSWAP gates in between
+    return [
+        (
+            kak.single_qubit_operations_after[0] @ kak.single_qubit_operations_before[0],
+            kak.single_qubit_operations_after[1] @ kak.single_qubit_operations_before[1],
+        )
+    ], kak.global_phase
+
+
+def _decomp_1sqrt_iswap_matrices(
+    kak: 'cirq.KakDecomposition',
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Returns the single-qubit matrices for the 1-SQRT_ISWAP decomposition.
+
+    Assumes canonical x, y, z and (x, y, z) = (π/8, π/8, 0) within tolerance.
+    """
+    return [  # Pairs of single-qubit unitaries, SQRT_ISWAP between each is implied
+        kak.single_qubit_operations_before,
+        kak.single_qubit_operations_after,
+    ], kak.global_phase
+
+
+def _decomp_2sqrt_iswap_matrices(
+    kak: 'cirq.KakDecomposition',
+    atol: float = 1e-8,
+) -> Tuple[Sequence[Tuple[np.ndarray, np.ndarray]], complex]:
+    """Returns the single-qubit matrices for the 2-SQRT_ISWAP decomposition.
+
+    Assumes canonical x, y, z and x >= y + |z| within tolerance.  For x, y, z
+    that violate this inequality, three sqrt-iSWAP gates are required.
+
+    References:
+        Towards ultra-high fidelity quantum operations: SQiSW gate as a native
+        two-qubit gate
+        https://arxiv.org/abs/2105.06074
+    """
+    x, y, z = kak.interaction_coefficients
+    a0, a1 = kak.single_qubit_operations_before
+    c0, c1 = kak.single_qubit_operations_after
+
+    # Computed gate parameters: Eq. 4, 6, 7, 8 of the paper
+    # range limits added for robustness to numerical error
+    def safe_arccos(v):
+        return np.arccos(np.clip(v, -1, 1))
+
+    def nonzero_sign(v):
+        return -1 if v < 0 else 1
+
+    _c = np.clip(
+        np.sin(x + y - z) * np.sin(x - y + z) * np.sin(-x - y - z) * np.sin(-x + y + z), 0, 1
+    )
+    alpha = safe_arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) + 2 * np.sqrt(_c))
+    beta = safe_arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) - 2 * np.sqrt(_c))
+    _4ccs = 4 * (np.cos(x) * np.cos(z) * np.sin(y)) ** 2  # Intermediate value
+    gamma = safe_arccos(
+        nonzero_sign(z)
+        * np.sqrt(_4ccs / (_4ccs + np.clip(np.cos(2 * x) * np.cos(2 * y) * np.cos(2 * z), 0, 1)))
+    )
+
+    # Inner single-qubit gates: Fig. 4 of the paper
```

Comment:
This is Fig 5. in the paper, also, gammas seem to be half angles and it seems you added a minus sign:
![image](https://user-images.githubusercontent.com/635073/124218727-cb0eef00-dac8-11eb-9e4c-2302d28440a6.png)

Reply:
This is true for alpha and beta as well (the angle used here is `-2 * t` instead of `t` in paper). I tried to find a counter case, but it looks like this works well. Though I'm also curious to find out the reason.

Reply:
The authors use a non-standard (-2x) definition of the rotation gates.  They confirmed this when I asked.

Reply:
I can add a comment.